jaxopt/_src/polyak_sgd.py

```diff
@@ -0,0 +1,185 @@
+
+"""SGD solver with Polyak step size."""
+
+from typing import Any
+from typing import Callable
+from typing import NamedTuple
+from typing import Optional
+from typing import Union
+
+import dataclasses
+
+import jax
+import jax.numpy as jnp
+
+from jaxopt import base
+from jaxopt.tree_util import tree_add
+from jaxopt.tree_util import tree_add_scalar_mul
+from jaxopt.tree_util import tree_l2_norm
+from jaxopt.tree_util import tree_scalar_mul
+from jaxopt.tree_util import tree_sub
+from jaxopt.tree_util import tree_zeros_like
+
+
+class PolyakSGDState(NamedTuple):
+  """Named tuple containing state information."""
+  iter_num: int
+  error: float
+  value: float
+  aux: Any
+  velocity: Optional[Any]
+
+
+@dataclasses.dataclass
+class PolyakSGD(base.IterativeSolverMixin, base.StochasticSolverMixin):
+  """SGD with Polyak step size.
+
+  This solver computes step sizes in an adaptive manner. If the computed step
+  size at a given iteration is smaller than ``max_step_size``, it is accepted.
+  Otherwise, ``max_step_size`` is used. This ensures that the solver does not
+  take over-confident steps. This is why ``max_step_size`` is the most important
+  hyper-parameter.
+
+  This implementation assumes that the interpolation property holds.
+
+  Attributes:
+    fun: a function of the form ``fun(params, *args, **kwargs)``, where
+      ``params`` are parameters of the model,
+      ``*args`` and ``**kwargs`` are additional arguments.
+    max_step_size: a maximum step size to use.
+    delta: a value to add in the denominator of the update (default: 0).
+    momentum: momentum strength, 0 corresponding to no momentum.
+    pre_update: a function to execute before the solver's update.
+      The function signature must be
+      ``params, state = pre_update(params, state, *args, **kwargs)``.
+    maxiter: maximum number of solver iterations.
+    tol: tolerance to use.
+    verbose: whether to print error on every iteration or not. verbose=True will
+      automatically disable jit.
+    implicit_diff: if True, enable implicit differentiation using cg,
+      if Callable, do implicit differentiation using callable as linear solver,
+      if False, use autodiff through the solver implementation (note:
+        this will unroll syntactic loops).
+    has_aux: whether ``fun`` outputs one (False) or more values (True).
+      When True it will be assumed by default that ``fun(...)[0]``
+      is the objective value. The auxiliary outputs are stored in
+      ``state.aux``.
+
+  References:
+    Berrada, Leonard and Zisserman, Andrew and Kumar, M Pawan.
+    "Training neural networks for and by interpolation".
+    International Conference on Machine Learning, 2020.
+
+    Loizou, Nicolas and Vaswani, Sharan and Laradji, Issam Hadj and
+    Lacoste-Julien, Simon.
+    "Stochastic polyak step-size for sgd: An adaptive learning rate for fast
+    convergence".
+    International Conference on Artificial Intelligence and Statistics, 2021.
```

Comment:
```suggestion
    International Conference on Artificial Intelligence and Statistics, 2021.
    https://arxiv.org/pdf/2002.10542.pdf
```

Reply:
(I just like when I can click in the ref and get the paper immediately :-) )

Reply:
Added, but I prefer pointing to the abstract page ;-)